app/src/org/commcare/utils/EncryptionUtils.java

```diff
@@ -1,17 +1,212 @@
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
 import java.security.MessageDigest;
 import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.UnrecoverableEntryException;
+import java.security.cert.CertificateException;
+import java.util.Random;
+
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.KeyGenerator;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.IvParameterSpec;
 
 /**
  * Utility class for encrypting submissions during the SaveToDiskTask.
  *
  * @author mitchellsundt@gmail.com
  */
+
 public class EncryptionUtils {
 
+    private static final int PASSPHRASE_LENGTH = 32;
+    private static final String SECRET_NAME = "secret";
+
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String ALGORITHM = KeyProperties.KEY_ALGORITHM_AES;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String BLOCK_MODE = KeyProperties.BLOCK_MODE_CBC;
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private static final String PADDING = KeyProperties.ENCRYPTION_PADDING_PKCS7;
+    private static KeyStore keystore = null;
+
+    //Gets the SecretKey from the Android KeyStore (creates a new one the first time)
+    private static SecretKey getKey()
+            throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException,
+            UnrecoverableEntryException, InvalidAlgorithmParameterException {
+        if(keystore == null) {
+            keystore = KeyStore.getInstance("AndroidKeyStore");
+            keystore.load(null);
+        }
+
+        KeyStore.Entry existingKey = keystore.getEntry(SECRET_NAME, null);
+        if (existingKey instanceof KeyStore.SecretKeyEntry) {
+            return ((KeyStore.SecretKeyEntry) existingKey).getSecretKey();
+        }
+
+        //Create key
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
+            KeyGenParameterSpec.Builder builder = new KeyGenParameterSpec.Builder(SECRET_NAME,
+                    KeyProperties.PURPOSE_ENCRYPT | KeyProperties.PURPOSE_DECRYPT);
+            builder.setBlockModes(BLOCK_MODE);
+            builder.setEncryptionPaddings(PADDING);
+            builder.setUserAuthenticationRequired(false);
+            builder.setRandomizedEncryptionRequired(true);
+
+            KeyGenerator generator = KeyGenerator.getInstance(ALGORITHM);
+            generator.init(builder.build());
+            return generator.generateKey();
+        }
+
+        return null;
+    }
+
+    //Generate a random passphrase
+    public static byte[] generatePassphrase() {
+        Random random;
+        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
```

Comment:
What is api check protecting here ? I see `SecureRandom` is available before Android O.

Reply:
It's the call to `getInstanceStrong()` that isn't supported in earlier versions. I'll add some code that attempts to create the `SecureRandom` a different way when possible (which includes specifying the algorithm).